keps/sig-architecture/1659-standard-topology-labels/README.md

```diff
@@ -0,0 +1,298 @@
+that standardizing on a small set of well-defined topology concepts will be a
+net win for the project at almost no cost to what users are actually doing with
+Kubernetes.
+
+[Envoy locality API]: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#envoy-v3-api-msg-config-core-v3-locality
+
+### Goals
+
+The goals of this KEP are to:
+   * build consensus that the two topology lables that ALREADY EXIST in Kubernetes are enough for most users
+   * determine whether a third level of topology is required or not
+   * produce short, descriptive, canonical documentation for theses labels
+
+### Non-Goals
+
+This KEP does NOT seek to:
+   * add new functionality that uses topology
+   * change existing functionality that uses topology
+   * solve the service topology problem
+
+## Proposal
+
+Kubernetes has always taken the position that "topology is arbitrary", and
+designs dealing with topology have had to take that into account.  Even so, the
+project has two commonly assumed labels - `topology.kubernetes.io/region` and
+`topology.kubernetes.io/zone` - which are used in many components, generally
+hard-coded and not extensible.  Those labels have relatively well understood
+meanings, and (so far) have been sufficient to represent what most people need.
+
+This KEP proposes to document those labels as "standard" and give them more
+rigorous definitions.  This also proposes that we discuss and decide whether a
+third level of topology is needed and if so, define it in the same manner as
+the existing labels.
+
+The resulting definitions should be specific enough that users and implementors
+understand what they mean, but not so rigid that they can not map them to the
+nearest constructs available in most environments.
+
+### Risks and Mitigations
+
+The primary risks here are:
+
+1) That we define these too loosely, such that users can not derive sufficient
+value from their use.
+
+2) That we define these too specifically, such that implementors can not use
+them to represent natural concepts in their environents.
+
+3) That we define these in a way that is incompatible with the ways they are
+alredy being used.
+
+4) That we preclude or design-out other uses of topology that users are using
+today.
+
+## Design Details
+
+### Reserve a label prefix
+
+Label prefixes allow us to group labels on common origin and meaning.  We
+propose to document somewhere (TBD) that the prefix "topology.kuberntes.io" is
+explicitly reserved for use in defining metadata about the physical or logical
+connectivity and grouping of Kubernetes nodes, and the associated behavioral
+and failure properties of those groups.
+
+This prefix is already in use.  This KEP just aims to formalize it.
+
+### Defining the meaning of existing labels
+
+This KEP proposes to define the meaning and semantics of the following labels:
+
+    * topology.kubernetes.io/region
+    * topology.kubernetes.io/zone
```

Comment:
Above, you say the definition of "region" and "zone" needs to address "the associated behavioral and failure properties of those groups". I think I agree. I'm interested in seeing the proposed definitions from folks in this area.

In particular, I'm curious how a single key-value label will be enough to describe the behaviours and failure properties. Would one alternative to this KEP (albeit a more invasive one) be to have actual `Region` and `Zone` objects in the API that could have additional properties associated with them?

Reply:
I don't see that as an alternative to, but rather a further extension.  If we assert that topology is region/zone/host then we may want to expose additional information about the possible values for those labels (network cost metric, etc)